File: ext/shapeop/libShapeOp/bindings/c++/runme.cpp
#include "Solver.h"
#include "Constraint.h"
#include <iostream>

void print_points(const ShapeOp::Matrix34 &p)
{
	for(int i = 0; i < 4; ++ i){
		std::cout << "Point " << i << " : ( ";
		ShapeOp::Vector3 current_pt = p.col(i);
		std::cout << std::scientific << current_pt.transpose();
		std::cout << " )" << std::endl;
	}
}

int main() {
  ShapeOp::Matrix34 p; //column major

  // The << operator reads elements row by row
  p << 0.0, 0.5, 0.5, 0.0,
       0.0, 0.0, 1.0, 1.0,
       0.0, 1.0, 0.0, 1.0;

  std::cout <<  "Input points:" << std::endl;
  print_points(p);

  ShapeOp::Solver s;
  s.setPoints(p);
  ShapeOp::Scalar weight = 1.0;
  //add a plane constraint to all the vertices.
  {
    std::vector<int> id_vector;
    id_vector.push_back(0); id_vector.push_back(1); id_vector.push_back(2); id_vector.push_back(3);
    auto c = std::make_shared<ShapeOp::PlaneConstraint>(id_vector, weight, s.getPoints());
    s.addConstraint(c);
  }
  //add a closeness constraint to the 1st vertex.
  {
    std::vector<int> id_vector;
    id_vector.push_back(0);
    auto c = std::make_shared<ShapeOp::ClosenessConstraint>(id_vector, weight, s.getPoints());
    s.addConstraint(c);
  }
  //add a closeness constraint to the 4th vertex.
  {
    std::vector<int> id_vector;
    id_vector.push_back(3);
    auto c = std::make_shared<ShapeOp::ClosenessConstraint>(id_vector, weight, s.getPoints());
    s.addConstraint(c);
  }
  //add a rigid constraint between 1st and 4th vertex.
  {
    std::vector<int> id_vector;
    id_vector.push_back(0); id_vector.push_back(3);
    auto c = std::make_shared<ShapeOp::SimilarityConstraint>(id_vector, weight, s.getPoints(),false);
    s.addConstraint(c);

  // edit the shapes one of which the rigid constraint brings the involved vertices close to.
    std::vector<ShapeOp::Matrix3X> shapes;
    ShapeOp::Matrix32 s1,s2; //column major

    s1 << 0.0, 0.5,
    	  0.0, 0.0,
		  0.0, 0.0;
    shapes.push_back(s1);

    s2 << 0.0, 1.0,
    	  0.0, 0.0,
		  0.0, 0.0;
    shapes.push_back(s2);

    c->setShapes(shapes);

  }
  s.initialize();
  s.solve(10);
  p = s.getPoints();
  std::cout << "Output points:" << std::endl;
  print_points(p);


  return 0;
}

File: ext/shapeop/libShapeOp/bindings/python/runme.py
#!/usr/bin/python
#/usr/local/bin/python3
import shapeopPython

#print array
def printArray(a, n):
  for c in range(0,3):
    for r in range(0,n):
      print a[r * 3 + c],
    print ""

def main():
  #create points with the swig doubleArray proxy
  #http://www.swig.org/Doc3.0/SWIGDocumentation.html#Library_carrays
  p = shapeopPython.doubleArray(12) #column major
  p[0] = 0.; p[3] = 0.5; p[6] = 0.5; p[9] = 0.;
  p[1] = 0.; p[4] = 0.; p[7] = 1.; p[10] = 1.;
  p[2] = 0.; p[5] = 1.; p[8] = 0.; p[11] = 1.;
  print "Input points:"
  printArray(p, 4)
  s=shapeopPython.shapeop_create()
  shapeopPython.shapeop_setPoints(s, p, 4)
  weight = 1.
  #add a plane constraint to all the vertices.
  ids = shapeopPython.intArray(4)
  ids[0] = 0; ids[1] = 1; ids[2] = 2; ids[3] = 3;
  shapeopPython.shapeop_addConstraint(s, "Plane", ids, 4, weight)
  #add a closeness constraint to the 1st vertex.
  ids = shapeopPython.intArray(1)
  ids[0] = 0;
  shapeopPython.shapeop_addConstraint(s, "Closeness", ids, 1, weight)
  #add a closeness constraint to the 4th vertex.
  ids = shapeopPython.intArray(1)
  ids[0] = 3;
  shapeopPython.shapeop_addConstraint(s, "Closeness", ids, 1, weight)
  shapeopPython.shapeop_init(s)
  shapeopPython.shapeop_solve(s, 10)
  shapeopPython.shapeop_getPoints(s, p, 4)
  shapeopPython.shapeop_delete(s)
  print "Output points:"
  printArray(p, 4)

if __name__ == "__main__":
  main()
